### DISCUSSION TITLE
Managing hierarchical data

### DISCUSSION THREAD

**Discussion by RenaudLN** [dagster-io/dagster#18878] [Q&A]
Hi all, my use case is the following:

* I create a hierarchical structure (directed graph) representing geographical areas (e.g. town -> county -> state -> country)
* I then leverage various datasets to estimate various datapoints at the lowest levels of the hierarchy, then aggregate those data to higher levels (e.g., population per town is summed to get the county population, then state, then country)
* The input hierarchical structure should be computed as a software defined asset and is dynamic

I was experimenting with partitions and partitions mapping which might solve my problem if the elements in my hierarchy were known in advance. Here is what I have so far as an explainer

```python
import random
from dagster import (
    AssetIn,
    Output,
    StaticPartitionsDefinition,
    StaticPartitionMapping,
    asset,
)

# Partitions
low_partition = StaticPartitionsDefinition(["000", "001", "002", "010", "011", "012"])
mid_partition = StaticPartitionsDefinition(["00", "01"])
top_partition = StaticPartitionsDefinition(["0"])

# Mappings
low_to_mid_mapping = StaticPartitionMapping(
    {"000": "00", "001": "00", "002": "00", "010": "01", "011": "01", "012": "01"}
)
mid_to_top_mapping = StaticPartitionMapping({"00": "0", "01": "0"})

@asset(partitions_def=low_partition)
def low_pop():
    population = random.randint(1, 11)
    return Output(population, metadata={"population": population})

@asset(
    partitions_def=mid_partition,
    ins={
        "low_pop": AssetIn(partition_mapping=low_to_mid_mapping),
    },
)
def mid_pop(low_pop):
    population = sum(low_pop.values())
    return Output(population, metadata={"population": population})

@asset(
    partitions_def=top_partition,
    ins={
        "mid_pop": AssetIn(partition_mapping=mid_to_top_mapping),
    },
)
def top_pop(mid_pop):
    population = sum(mid_pop.values())
    return Output(population, metadata={"population": population})
```

![example-hierarchichal](https://github.com/dagster-io/dagster/assets/20662281/ffc50cda-16ac-4870-84d9-f72aada4d821)

* Can I create those assets dynamically based on my hierarchy asset?
* Can I create those partitions and mappings dynamically based on my hierarchy asset?

Disclaimer: just starting with Dagster 🙂

**Comment by RenaudLN** [comment]
This dynamic graph creation works well in Apache Airflow if anyone has an idea of how to convert it to Dagster it'd be greatly appreciated 🙂 

```python
import random

import networkx as nx
from airflow import DAG
from airflow.decorators import task


ASSET_HIERARCHY = ["township", "county", "state"]


def save_population(asset: str, population: int):
    with open(f"/home/user/population/{asset}", "w") as fp:
        fp.write(str(population))

def load_population(asset: str):
    with open(f"/home/user/population/{asset}", "r") as fp:
        return int(fp.read())

def get_topo():
    topo = nx.DiGraph()
    topo.add_nodes_from(
        [
            ("0", {"level": "state"}),
            ("00", {"level": "county"}),
            ("000", {"level": "township"}),
            ("001", {"level": "township"}),
            ("002", {"level": "township"}),
            ("01", {"level": "county"}),
            ("010", {"level": "township"}),
            ("011", {"level": "township"}),
            ("02", {"level": "county"}),
            ("020", {"level": "township"}),
            ("021", {"level": "township"}),
        ]
    )
    topo.add_edges_from(
        [
            ("0", "00"), ("0", "01"), ("0", "02"), ("00", "000"), ("00", "001"), ("00", "002"),
            ("01", "010"), ("01", "011"), ("02", "020"), ("02", "021"),
        ]
    )
    return topo

with DAG(dag_id="population_uptake") as dag:

    @task
    def get_level(level):
        topo = get_topo()
        return sorted(a for a in topo if topo.nodes[a]["level"] == level)

    @task
    def get_children_kwargs(asset):
        topo = get_topo()
        return {"asset": asset, "children": sorted(topo.successors(asset))}

    @task
    def get_lower_level_population(asset: str):
        population = random.randint(1, 11)
        save_population(asset, population)
        return population

    @task
    def aggregate_population(asset: str, children: list[str]):
        population = 0
        for child in children:
            population += load_population(child)
        save_population(asset, population)
        return population

    pops = []
    for i, level in enumerate(ASSET_HIERARCHY):
        assets = get_level(level)
        if i == 0:
            pop = get_lower_level_population.expand(asset=assets)
        else:
            children_kwargs = get_children_kwargs.expand(asset=assets)
            pop = aggregate_population.expand_kwargs(children_kwargs)
        pops.append(pop)

    for pop0, pop1 in zip(pops[:-1], pops[1:]):
        pop0 >> pop1
```

![image](https://github.com/dagster-io/dagster/assets/20662281/e6fb2d6d-66ef-4c2a-913b-22cc8ce79208)

**Comment by sryza** [comment]
> Can I create those assets dynamically based on my hierarchy asset?

Hey @RenaudLN - what exactly do you mean by dynamically here? I.e. in what situations do you want it to get updated? In the implementation you included with `StaticPartitionsDefinition`s, it will get updated whenever you reload your code location.  If that works for you, then I think this is a good way to approach it.

With Airflow, I think you're getting basically the same behavior (i.e. when Airflow reloads your DAG).

**Reply by RenaudLN** [reply]
In my usecase, the topology (nodes and mapping between levels) should be defined as part of the DAG, which will then impact all the downstream pipeline tasks.

I see that I can create some dynamic partitions with Dagster so maybe a sensor on the topology asset could populate the partitions properly?

Something that I still have no idea how to do with Dagster is how to dynamically map my dynamic partitions between levels.

**Reply by sryza** [reply]
You're saying that you have an asset that defines the topology? So in your example above, the `def get_topo():` function would ideally be reading from a database instead of from a hardcoded list?

A sensor with dynamic partitions could work for this. One limitation however is that we don't currently have rich support for custom partition mappings for dynamic partitions: https://github.com/dagster-io/dagster/issues/13139.

**Reply by RenaudLN** [reply]
Yes that's exactly right. Thanks for pointing me to this issue, I'll follow the progress there 🙂
